### DISCUSSION TITLE
How to raycast occupancy?

### DISCUSSION THREAD

**Discussion by Divelix** [ethz-asl/wavemap#80] [Q&A] [answered]
My idea is to extract depth maps from octree map given arbitrary camera pose and intrinsics. As I understand, core functionality to accomplish this is to be able to cast ray and get coordinates of the first point whose occupancy is higher than some threshold. 

So my question is where to start? I found [this Ray class](https://github.com/ethz-asl/wavemap/blob/main/library/cpp/include/wavemap/core/utils/iterate/ray_iterator.h), but not even sure if this is the right thing I need)

P.S. congratulations on the great refactoring, new project structure is awesome!

**Comment by Divelix** [comment]
First idea is to sample points along the ray and call `interpolate` for each point. Not the most efficient way to accomplish my goal, but it can be implemented on python side (which is a huge plus for me).

**Comment by victorreijgwart** [comment, marked as answer]
Thanks! I'm glad to hear you like the new structure. It took some time, but it should solve all the issues you highlighted in #41 :)

Ray tracing is computationally expensive, so moving critical loops to C++ to improve performance is probably worth it. You could work primarily in Python and, when needed, migrate heavy sections directly into the pywavemap interface (this is probably easier than moving them to the C++ library and then writing bindings separately). With an [Editable install](https://ethz-asl.github.io/wavemap/pages/installation/python.html#editable-install) and `-Ceditable.rebuild=true`, any changes to the C++ code will automatically recompile when running your Python script.

You can leverage the `RayIterator` to visit all voxels intersected by a ray, ordered by distance, as follows:

```cpp
const Point3D start_point; // the start point of your ray (usually the sensor's origin)
const Point3D end_point;   // the end point of your ray (usually at a given max dist from your sensor)
const FloatingPoint occupancy_threshold = 0.1f; // threshold above which you consider a cell occupied
const FloatingPoint min_cell_width = map.getMinCellWidth();
const Ray ray(start_point, end_point, min_cell_width);
for (const Index3D& ray_voxel_index : ray) {
  if (occupancy_threshold < map.getCellValue(ray_voxel_index)) {
    const Point3D voxel_center = convert::indexToCenterPoint(ray_voxel_index, min_cell_width);
    return (voxel_center - start_point).norm();
  }
}
return (end_point - start_point).norm();
```
This approach ensures that you visit every voxel intersected by the ray, whereas query points found by interpolating the ray are likely to miss some voxels. Another advantage is that this way, it's easy to end the queries as soon as the nearest occupied voxel is found.

**Reply by victorreijgwart** [reply]
In case you want to render an image for a virtual sensor whose projection model matches one of the models implemented in wavemap, you can also use the following code to generate the rays:

```cpp
  const Transformation3D T_W_C;  // The pose of your sensor
  const Point3D start_point = T_W_C.getPosition();
  const FloatingPoint max_range;  // Maximum ray tracing ray length

  for (const Index2D& index :
       Grid<2>(Index2D::Zero(), 
                      depth_image.getDimensions() - Index2D::Ones())) {
    const Vector2D image_xy = projection_model.indexToImage(index);
    const Point3D C_point =
        projection_model.sensorToCartesian(image_xy, max_range);

    const Point3D end_point = T_W_C * C_point;
    // Call the ray tracing routine to trace from start_point to end_point
    // Store the resulting distance in depth_image.at(index);
  }
```

**Reply by victorreijgwart** [reply]
Inside the ray trace loop, I'd also recommend using the [query accelerator](https://ethz-asl.github.io/wavemap/pages/tutorials/cpp.html#accelerators) which should give a very big speedup for this access pattern. It's not implemented for [HashedChunkedWaveletOctrees](https://ethz-asl.github.io/wavemap/pages/parameters/map.html#hashed-chunked-wavelet-octree) yet though, so you'd have to switch the map type to a [HashedWaveletOctree](https://ethz-asl.github.io/wavemap/pages/parameters/map.html#hashed-wavelet-octree).

**Reply by Divelix** [reply]
It works, thank you! Given image of size 1280x720: my python crutch described above took ~3 and ~4 minutes for nearest and trilinear interpolations respectively, while your raycast takes 17 seconds (and depth quality is much better).
I could make PR if you consider these things useful.
Images are: nearest, trilinear and raycast variants taken from the same pose on the same map.
| nearest | trilinear | raycast |
|--------|--------|--------|
| ![wm_nearest](https://github.com/user-attachments/assets/556b386a-d034-4783-9a99-78634bd5b4c5)  | ![wm_trilinear](https://github.com/user-attachments/assets/30507503-0a47-47e5-a90d-5525c70a75e9)  | ![wm_raycast](https://github.com/user-attachments/assets/5a417390-8f7a-423e-8473-ac0eae4df9f5) |

Map resolution is 5cm here.

**Reply by Divelix** [reply]
Moving depth image loop to C++ side with query accelerator should make it even faster, I just haven't figured it out yet.

**Reply by Divelix** [reply]
Well, plain moving pixels iteration loop into C++ gave no gain in speed, while applying `QueryAccelerator` decreased compute wall time from 17 to 2 seconds for 1280x720 image!

**Reply by victorreijgwart** [reply]
That's great! It would also be straightforward to add support for multithreading, using the tooling we already implemented for multi-threaded measurement integration, etc. Would this be useful for you? If so, I'll add it to the PR you already created (thanks for doing so!).

**Reply by Divelix** [reply]
Of course it would be useful (also I'm curious how multi threaded version would look like in code)! I'm glad you like the PR, feel free to edit it.

**Reply by victorreijgwart** [reply]
I added some basic multi-threaded implementation [here](https://github.com/Divelix/wavemap/blob/raycast/library/python/src/raycast.cc#L80-L116). Does it work for you? When I render the image, it looks the same with or without multi-threading. But in general, the image didn't show much. It might just be that I didn't find a nice camera viewpoint for the map of panoptic mapping's flat. Do you maybe have some camera pose(s) you recommend for testing?

**Reply by Divelix** [reply]
Multi-threaded version gives ~x2 speedup for my case (it was 2.3 seconds, now it is 1.3). Resulting image looks fine.

**Reply by Divelix** [reply]
The thing is, I created map from my own data and for testing I just use pose of the first frame (identity pose).

**Reply by victorreijgwart** [reply]
Thanks for testing it, and good idea to just use poses from input frames. I'll implement tiling to see if that improves performance a bit more. After that, I'll start preparing the merge.

**Reply by victorreijgwart** [reply]
@Divelix, running the parallel ray casting jobs over image tiles (rather than creating one job per pixel) seems to further improve performance (~2 fps on my laptop), so I’ve pushed the changes. Let me know if it works on your end as well. I'll start preparing the merge, though it may take a bit of time as we have a few deadlines this week.

**Reply by Divelix** [reply]
@victorreijgwart I just tried it: now it takes ~180ms on my machine, wow...

**Reply by victorreijgwart** [reply]
@Divelix, the rendering feature is almost ready for release (PR #89). I generalized the code a bit so that it can also be used to render depth images for other sensor types (e.g., LiDARs), and it now reuses the same thread pool for subsequent render jobs. This should make it more efficient to render large sets of images. Could you give this version a try to see if it works well for you and if the performance is still good? The updated example script is [here](https://github.com/ethz-asl/wavemap/blob/feature/depth_image_renderer/examples/python/render/raycast_depth_image.py).

**Reply by Divelix** [reply]
@victorreijgwart sorry for delay, I just checked: new version took ~130ms in my setup. So now it is even faster!
